aten/src/ATen/native/sparse/SparseBinaryOpIntersectionCommon.h

```diff
@@ -270,14 +270,11 @@ void _sparse_binary_op_intersection_kernel_impl(
     );
     auto strides = contiguous_strides(broadcasted_sparse_dim_shape);
     auto strides_len = static_cast<int64_t>(strides.size());
-    auto hash_coeffs = at::empty(
+    auto hash_coeffs = at::from_blob(
+        strides.data(),
         {strides_len},
-        probably_coalesced._indices().options().device(kCPU).dtype(kHash));
-    // Copy with a potential casting. Is there a nicer way?
-    for (const auto i : c10::irange(strides_len)) {
-      hash_coeffs[i] = strides[i];
-    }
-    hash_coeffs = hash_coeffs.to(probably_coalesced.device());
+        probably_coalesced._indices().options().device(kCPU).dtype(kLong));
+    hash_coeffs = hash_coeffs.to(probably_coalesced.device(), kHash);
```

Comment:
For even more performance we might want to consider using asynchronous transfers. That'll however require a move to pinned memory. 

Ideally allocating GPU memory is entirely avoided. Can you collapse dimensions and create a ceiling on dim? You could also template on some reasonable ceiling value (say support up to 8 dimensional Tensors efficiently) and then fall back to allocating GPU memory. This is not unlike the SmallVector approach that avoids heap allocation on CPU.

Reply:
Makes sense, will, so that the whole kernel is sync-less.

Reply:
Well, these are sparse dimension, so we cannot trivially collapse them. As a potential workaround, we could `std::array` as a backbone for some predefined dim sizes. What do you think?

Reply:
It is not that trivial since these dimensions are sparse. I have put on a threshold on when to perform the sync. I suppose having 10 sparse dims is sufficient in most cases? Looks like CUDA does not complain at all about statics arrays of length <= 64.